### PR TITLE
task/WG-245: Adding core-wrappers library

### PR DESCRIPTION
**Pull request by tjgrafft**
## Overview: ##
Copied TUP's core-wrappers library into the hazmapper-react application. This was the result of our meeting with Wes regarding styling and layout. We already have tacc's core-styles, but adding wrappers in addition to the core-components was necessary. 
## PR Status: ##

* [x] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [WG-245](https://tacc-main.atlassian.net/browse/WG-245)

## Summary of Changes: ##

- Added line to tell eslinting to ignore new core-wrappers directory
- Added code to tell jest to ignore running test of core-wrappers directory (probably need to test these down the road)
- Added react-dropzone package (requirement for wrapper)
- Added react-step-wizard package (requirement for wizard wrapper)
- Copied over tup's core-wrappers files from their repo (this is a short-term solution until the design team is able to create a npm package we can use to install the wrappers that way)
- Created a simple component using FieldWrapperFormik wrapper to showcase that new core-wrappers are working with changes. (see screenshot below) WG-246 will have a more complex and working FieldWrapperFormik modal. 

## Testing Steps: ##
1. Run npm ci command
2. Go to localhost:4200
3. Click the Create Map button on Main menu
4. Compare modal layout with screenshot below
5. Note: I'll get rid of MapModalExample, the changes in MainMenu and the changes in index.css before this is merged to main--as they're not needed and only for yall to test with for this PR. 

## UI Photos:
Screenshot 2024-02-07 at 10 48 47 AM

### CODE REVIEW

File: react/jest.config.cjs [hunk 1]
```diff
@@ -179,7 +177,7 @@ module.exports = {
   // timers: "real",
 
   // A map from regular expressions to paths to transformers
-  transform: {'^.+\\.(js|jsx)?$': 'babel-jest'},
```
Comment (nathanfranklin) [inline react/jest.config.cjs:182]: is this change needed?

File: react/.eslintrc.json [hunk 2]
```diff
@@ -1,32 +1,32 @@
...
-	"prettier/prettier": ["error"],
-        "@typescript-eslint/no-unused-vars": "error",
-        "@typescript-eslint/no-explicit-any": "off"
-    },
-    "settings": {
-        "react": {
-            "version": "detect"
-        }
-    }
+  "env": {
```
Comment (nathanfranklin) [inline react/.eslintrc.json:2]: this just has just shifts things 2 spaces to left.  so first level has 2-spaces indention and then subsequent levels have 4 spaces indention. 

 i think its best we avoid changing linting/indentions on top level things to make reviews easier until we do https://tacc-main.atlassian.net/browse/WG-248.

File: react/jest.config.cjs (hunk 1 above)
Comment (tjgrafft) [reply react/jest.config.cjs:182]: The unit test for MainMenu was failing with a TypeError: Cannot read properties of undefined (reading 'createContext') error before I added that transformation for the ts and tsx file type. React was already being imported in MainMenu too, but not being read properly before apparently

**Comment by nathanfranklin** [comment]
LGTM.

I left two minor comments to consider. 

> Note: I'll get rid of MapModalExample, the changes in MainMenu and the changes in index.css before this is merged to main--as they're not needed and only for yall to test with for this PR.

Can you ping me again for final review ✅

File: react/jest.config.cjs (hunk 1 above)
Comment (wesleyboar) [reply react/jest.config.cjs:182]: Perhaps that bug occurred because of the (beneficial, I think) change of —

```diff
-        "test": "jest --testPathIgnorePatterns='<rootDir>/src/core-components/'",
+        "test": "jest",
```

— in `react/package.json`. Instead of ignoring, they must… now be transformed?

Comment (wesleyboar) [reply react/jest.config.cjs:182]: Oh, I maybe morning-muddle-minded. I read the full PR description and more of the diff, now. There is other possibly relevant code.

Comment (tjgrafft) [reply react/jest.config.cjs:182]: Yeah MainMenu doesn't fall under the core-components umbrella, and I just moved the testPathIgnorePatterns line to the jest.config file instead. I'm not quite sure why ts and tsx file need to be transformed now, but weren't before. But looking at the diff of package-lock.json there's some babel items that were upgraded, so it could be related to that

File: react/jest.config.cjs [hunk 3]
```diff
@@ -28,9 +28,7 @@ module.exports = {
   coverageDirectory: 'coverage',
 
   // An array of regexp pattern strings used to skip coverage collection
-  coveragePathIgnorePatterns: [
-    "src/core-components/"
-  ],
+  coveragePathIgnorePatterns: ['src/core-components/', 'src/core-wrappers'],
```
Comment (wesleyboar) [inline react/jest.config.cjs:31]: I ran `npm ci` and `npm run test` with this new `coveragePathIgnorePatterns`, and with an empty list. Both produced the same results. Does that mean tests are passing for `src/core-…`? Should tests then be allowed ot run? Am I misunderstanding?

<details><summary><code>  coveragePathIgnorePatterns: ['src/core-components/', 'src/core-wrappers'],</code></summary>

```
...
----------------------------|---------|----------|---------|---------|-------------------
Test Suites: 3 failed, 4 passed, 7 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        16.494 s
Ran all test suites.
```

</details>
<details><summary><code>  coveragePathIgnorePatterns: [],</code></summary>

```
...
----------------|---------|----------|---------|---------|-------------------
Test Suites: 3 failed, 4 passed, 7 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        14.151 s
Ran all test suites.
```

</details>

**Review by nathanfranklin — APPROVED** [review]
LGTM 👍 

just resolve https://github.com/TACC-Cloud/hazmapper/pull/204/files#r1484562505 before merging

File: react/.eslintignore [hunk 4]
```diff
@@ -1,4 +1,5 @@
 dist/
 build/
 coverage/
-src/core-components/
\ No newline at end of file
+src/core-components/
+src/core-wrappers/
```
Comment (wesleyboar) [inline react/.eslintignore]: This change is effectual. There are 15 lint problems in the new code. Acceptable.[^1]

<details><summary>Why didn't tup-ui fix these?</summary>

I never saw these lint issues on tup-ui via `nx format:check`, because… oh my gosh, [`libs/core-components/.eslintrc.json`](https://github.com/TACC/tup-ui/blob/v1.1.5/libs/core-components/.eslintrc.json#L3) and [`libs/core-wrappers/.eslintrc.json`](https://github.com/TACC/tup-ui/blob/v1.1.5/libs/core-wrappers/.eslintrc.json#L3) ignore everything too.

</details>

<details><summary>Linting the new <code>src/core-…</code> results in 15 problems.</summary>

```
/Users/wbomar/Code/hazmapper/react/src/core-components/Button/Button.test.jsx
  11:54  error  'getByTestId' is defined but never used  @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-components/DropdownSelector/DropdownSelector.jsx
  26:50  error  'className' is missing in props validation  react/prop-types

/Users/wbomar/Code/hazmapper/react/src/core-components/InfiniteScrollTable/InfiniteScrollTable.jsx
   84:11  error  Missing "key" prop for element in iterator  react/jsx-key
   86:15  error  Missing "key" prop for element in iterator  react/jsx-key
   95:13  error  Missing "key" prop for element in iterator  react/jsx-key
   98:19  error  Missing "key" prop for element in iterator  react/jsx-key
  129:22  error  'offset' is defined but never used          @typescript-eslint/no-unused-vars
  133:17  error  'row' is defined but never used             @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-components/SectionHeader/SectionHeader.jsx
  45:3  error  'isNestedHeader' is missing in props validation  react/prop-types

/Users/wbomar/Code/hazmapper/react/src/core-components/Sidebar/Sidebar.jsx
  15:65  error  'hidden' is defined but never used  @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-wrappers/FieldWrapperFormik/fields/FormikCheck.tsx
  10:3  error  'required' is defined but never used  @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-wrappers/FieldWrapperFormik/fields/FormikFileInput/FileDropzone.tsx
  81:37  error  'i' is defined but never used  @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-wrappers/Navbar/Navbar.tsx
  36:18  error  'end' is defined but never used  @typescript-eslint/no-unused-vars

/Users/wbomar/Code/hazmapper/react/src/core-wrappers/utils/withBuilder.tsx
  20:11  error  'slice' is defined but never used  @typescript-eslint/no-unused-vars
  21:11  error  'slice' is defined but never used  @typescript-eslint/no-unused-vars

✖ 15 problems (15 errors, 0 warnings)
```

</details>

[^1]: Several lint problems seem easy to fix. But, given that no one has been linting them thus far, I can't strongly argue to fix them immediately. It seems a `@tacc/core-…` package responsibility to lint it's own code.

**Review by wesleyboar — COMMENTED** [review]
Ignoring **lint** is reasonable, because the new code introduces problems the coming Node package shoudl fix.

Ignoring **tests** seems unnecessary, because it adds code that will be removed later, and the tests pass regardless.

Feedback welcome.

File: react/jest.config.cjs (hunk 3 above)
Comment (wesleyboar) [reply react/jest.config.cjs:31]: One can argue "ignore lint, cuz it is responsibility of future package; thus ignore tests for the same reason". But if the tests pass, this "ignore" code seems destined to become cruft.

Comment (tjgrafft) [reply react/jest.config.cjs:31]: coveragePathIgnorePatterns should be related to the coverage checks for those dir, while testPathIgnorePatterns is ignoring the actual tests for those directories. Those 7 tests suites and 4 tests you got with npm run test are all outside those directories, so I'm surprised 3 test suites failed for you. Don't think that happened for Nathan and me. Hmm. Might try removing node modules and the package-lock and running npm install and then run the tests again and see if those 3 still fail?

Comment (wesleyboar) [reply react/jest.config.cjs:31]: Thanks. Heh, I had not distinguished coverage and test. 😅

Note: I ran —

1. `cd react`
2. `rm -rf node_modules`
3. `npm ci`
4. `npm run test`

— then —

1. `rm -rf node_modules`
2. `rm package-lock.json`
3. `npm install`
4. `npm run test`

<details><summary>Same output each time.</summary>

```
Test Suites: 3 failed, 4 passed, 7 total
Tests:       4 passed, 4 total
Snapshots:   0 total
```

</details>

Comment (tjgrafft) [reply react/jest.config.cjs:31]: When you get the chance, do you mind sharing which 3 are failing for you and any error output those 3 tests produce that could hint at the issue

Comment (wesleyboar) [reply react/jest.config.cjs:31]: Sure.

<details><summary>Output</summary>

```
➜  react git:(master) npm run test

> hazmapper@0.0.0 test
> jest


 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
 RUNS  src/components/Map/Map.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
  console.warn
    Warning: `fetch` is not available. Please supply a custom `fetchFn` property to use `fetchBaseQuery` on SSR environments.

       9 | export const geoapi = createApi({
      10 |   reducerPath: 'api',
    > 11 |   baseQuery: fetchBaseQuery({
         |                            ^
      12 |     baseUrl: BASE_URL,
      13 |     prepareHeaders: (headers, api) => {
      14 |       // TODO check if logged in as we don't want to add if public

      at fetchBaseQuery (node_modules/@reduxjs/toolkit/dist/query/rtk-query.cjs.development.js:363:13)
      at Object.<anonymous> (src/redux/api/geoapi.ts:11:28)
      at Object.require (src/redux/reducers/reducers.ts:2:1)
      at Object.require (src/redux/store.ts:2:1)
      at Object.require (src/requests.ts:2:1)
      at Object.<anonymous> (src/requests.test.ts:1:1)


 RUNS  src/pages/Callback/Callback.test.tsx
  console.warn
    Warning: `fetch` is not available. Please supply a custom `fetchFn` property to use `fetchBaseQuery` on SSR environments.

       9 | export const geoapi = createApi({
      10 |   reducerPath: 'api',
    > 11 |   baseQuery: fetchBaseQuery({
         |                            ^
      12 |     baseUrl: BASE_URL,
      13 |     prepareHeaders: (headers, api) => {
      14 |       // TODO check if logged in as we don't want to add if public

      at fetchBaseQuery (node_modules/@reduxjs/toolkit/dist/query/rtk-query.cjs.development.js:363:13)
      at Object.<anonymous> (src/redux/api/geoapi.ts:11:28)
      at Object.require (src/redux/reducers/reducers.ts:2:1)
      at Object.require (src/redux/store.ts:2:1)
      at Object.require (src/requests.ts:2:1)
      at Object.require (src/hooks/projects/useProjects.ts:3:1)
      at Object.require (src/hooks/index.ts:1:1)
      at Object.require (src/pages/MainMenu/MainMenu.tsx:8:1)
      at Object.<anonymous> (src/pages/MainMenu/MainMenu.test.tsx:3:1)


 RUNS  src/pages/Callback/Callback.test.tsx

 RUNS  src/pages/Callback/Callback.test.tsx
  console.warn
    Warning: `fetch` is not available. Please supply a custom `fetchFn` property to use `fetchBaseQuery` on SSR environments.

       9 | export const geoapi = createApi({
      10 |   reducerPath: 'api',
    > 11 |   baseQuery: fetchBaseQuery({
         |                            ^
      12 |     baseUrl: BASE_URL,
      13 |     prepareHeaders: (headers, api) => {
      14 |       // TODO check if logged in as we don't want to add if public

      at fetchBaseQuery (node_modules/@reduxjs/toolkit/dist/query/rtk-query.cjs.development.js:363:13)
      at Object.<anonymous> (src/redux/api/geoapi.ts:11:28)
      at Object.require (src/redux/reducers/reducers.ts:2:1)
      at Object.require (src/redux/store.ts:2:1)
      at Object.<anonymous> (src/pages/Callback/Callback.test.tsx:5:1)


 RUNS  src/pages/Callback/Callback.test.tsx
 PASS  src/pages/StreetviewCallback/StreetviewCallback.test.tsx (12.653 s)

 RUNS  src/pages/Callback/Callback.test.tsx
  ✓ renders streetview callback (63 ms)

 RUNS  src/pages/Callback/Callback.test.tsx


 RUNS  src/pages/Callback/Callback.test.tsx
 PASS  src/pages/Callback/Callback.test.tsx (12.72 s)

 RUNS  src/pages/Callback/Callback.test.tsx
  ✓ renders callback (69 ms)

 RUNS  src/pages/Callback/Callback.test.tsx


 RUNS  src/pages/Callback/Callback.test.tsx
  console.warn
    Warning: `fetch` is not available. Please supply a custom `fetchFn` property to use `fetchBaseQuery` on SSR environments.

       9 | export const geoapi = createApi({
      10 |   reducerPath: 'api',
    > 11 |   baseQuery: fetchBaseQuery({
         |                            ^
      12 |     baseUrl: BASE_URL,
      13 |     prepareHeaders: (headers, api) => {
      14 |       // TODO check if logged in as we don't want to add if public

      at fetchBaseQuery (node_modules/@reduxjs/toolkit/dist/query/rtk-query.cjs.development.js:363:13)
      at Object.<anonymous> (src/redux/api/geoapi.ts:11:28)
      at Object.require (src/redux/reducers/reducers.ts:2:1)
      at Object.require (src/redux/store.ts:2:1)
      at Object.require (src/requests.ts:2:1)
      at Object.require (src/hooks/projects/useProjects.ts:3:1)
      at Object.require (src/hooks/index.ts:1:1)
      at Object.require (src/pages/Login/Login.tsx:6:1)
      at Object.<anonymous> (src/pages/Login/Login.test.tsx:3:1)

 PASS  src/components/Map/Map.test.tsx (12.921 s)
  ✓ renders map (424 ms)

  console.warn
    Warning: `fetch` is not available. Please supply a custom `fetchFn` property to use `fetchBaseQuery` on SSR environments.

       9 | export const geoapi = createApi({
      10 |   reducerPath: 'api',
    > 11 |   baseQuery: fetchBaseQuery({
         |                            ^
      12 |     baseUrl: BASE_URL,
      13 |     prepareHeaders: (headers, api) => {
      14 |       // TODO check if logged in as we don't want to add if public

      at fetchBaseQuery (node_modules/@reduxjs/toolkit/dist/query/rtk-query.cjs.development.js:363:13)
      at Object.<anonymous> (src/redux/api/geoapi.ts:11:28)
      at Object.require (src/redux/authSlice.ts:8:1)
      at Object.require (src/pages/Logout/Logout.tsx:4:1)
      at Object.<anonymous> (src/pages/Logout/Logout.test.tsx:3:1)

  console.warn
    Detected an action creator with type "auth/logout" being dispatched. 
    Make sure you're calling the action creator before dispatching, i.e. `dispatch(logout())` instead of `dispatch(logout)`. This is necessary even if the action has no payload.

      13 |
      14 |   useEffect(() => {
    > 15 |     dispatch(logout);
         |     ^
      16 |   }, [dispatch]);
      17 |
      18 |   return (

      at node_modules/@reduxjs/toolkit/src/immutableStateInvariantMiddleware.ts:160:31
      at dispatch (src/pages/Logout/Logout.tsx:15:5)
      at commitHookEffectListMount (node_modules/react-dom/cjs/react-dom.development.js:3926:112)
      at commitPassiveMountOnFiber (node_modules/react-dom/cjs/react-dom.development.js:4139:1573)
      at commitPassiveMountEffects_complete (node_modules/react-dom/cjs/react-dom.development.js:4139:869)
      at commitPassiveMountEffects_begin (node_modules/react-dom/cjs/react-dom.development.js:4139:570)
      at commitPassiveMountEffects (node_modules/react-dom/cjs/react-dom.development.js:4139:196)
      at flushPassiveEffectsImpl (node_modules/react-dom/cjs/react-dom.development.js:4636:402)
      at flushPassiveEffects (node_modules/react-dom/cjs/react-dom.development.js:4631:371)
      at node_modules/react-dom/cjs/react-dom.development.js:4566:84
      at flushActQueue (node_modules/react/cjs/react.development.js:2332:28)
      at act (node_modules/react/cjs/react.development.js:2256:15)
      at node_modules/@testing-library/react/dist/act-compat.js:87:25
      at renderRoot (node_modules/@testing-library/react/dist/pure.js:171:26)
      at render (node_modules/@testing-library/react/dist/pure.js:256:10)
      at Object.<anonymous> (src/pages/Logout/Logout.test.tsx:9:31)

 PASS  src/pages/Logout/Logout.test.tsx (13.269 s)
  ✓ renders logout (283 ms)

 FAIL  src/requests.test.ts
  ● Test suite failed to run

    src/hooks/environment/useAppConfiguration.ts:4:47 - error TS2307: Cannot find module '../../secret_local' or its corresponding type declarations.

    4 import { localDevelopmentConfiguration } from '../../secret_local';
                                                    ~~~~~~~~~~~~~~~~~~~~

 FAIL  src/pages/MainMenu/MainMenu.test.tsx
  ● Test suite failed to run

    src/hooks/environment/useAppConfiguration.ts:4:47 - error TS2307: Cannot find module '../../secret_local' or its corresponding type declarations.

    4 import { localDevelopmentConfiguration } from '../../secret_local';
                                                    ~~~~~~~~~~~~~~~~~~~~

 FAIL  src/pages/Login/Login.test.tsx
  ● Test suite failed to run

    src/hooks/environment/useAppConfiguration.ts:4:47 - error TS2307: Cannot find module '../../secret_local' or its corresponding type declarations.

    4 import { localDevelopmentConfiguration } from '../../secret_local';
                                                    ~~~~~~~~~~~~~~~~~~~~

----------------------------|---------|----------|---------|---------|-------------------
File                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------------------|---------|----------|---------|---------|-------------------
All files                   |   70.73 |    27.27 |   51.72 |   70.73 |                   
 __fixtures__               |     100 |      100 |     100 |     100 |                   
  featuresFixture.ts        |     100 |      100 |     100 |     100 |                   
  tileServerLayerFixture.ts |     100 |      100 |     100 |     100 |                   
 components/Map             |   88.88 |      100 |   71.42 |   88.88 |                   
  Map.tsx                   |   88.88 |      100 |   71.42 |   88.88 | 30,70             
 pages/Callback             |      75 |       50 |     100 |      75 |                   
  Callback.tsx              |      75 |       50 |     100 |      75 | 19-23,31-34       
 pages/Logout               |   91.66 |      100 |   66.66 |   91.66 |                   
  Logout.tsx                |   91.66 |      100 |   66.66 |   91.66 | 11                
 pages/StreetviewCallback   |     100 |      100 |     100 |     100 |                   
  StreetviewCallback.tsx    |     100 |      100 |     100 |     100 |                   
 redux                      |   67.85 |      100 |    37.5 |   67.85 |                   
  authSlice.ts              |      50 |      100 |      20 |      50 | 25-41,48-52       
  projectsSlice.ts          |   83.33 |      100 |      50 |   83.33 | 13                
  store.ts                  |     100 |      100 |     100 |     100 |                   
 redux/api                  |   41.66 |        0 |      25 |   41.66 |                   
  geoapi.ts                 |   41.66 |        0 |      25 |   41.66 | 15-27,33-37       
 redux/reducers             |     100 |      100 |     100 |     100 |                   
  reducers.ts               |     100 |      100 |     100 |     100 |                   
 utils                      |      45 |        0 |      25 |      45 |                   
  authUtils.ts              |      45 |        0 |      25 |      45 | 6-14,22-26,32,36  
----------------------------|---------|----------|---------|---------|-------------------
Test Suites: 3 failed, 4 passed, 7 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        16.61 s
Ran all test suites.
```

</details>

_P.S. The `package-lock.json` did not change upon removal and `npm install`, so I think it is not the culprit._

Comment (tjgrafft) [reply react/jest.config.cjs:31]: Ah I see the issue. To run locally with our project you need to create a secret_local.ts file (it gets ignored by git since it's got a secret JWT in it). Then put your JWT in it. We have an example in the project called secrets_local.example.ts. So the tests are failing for you bc my useAuthenticated hook needs an authenticated user to work.

Comment (wesleyboar) [reply react/jest.config.cjs:31]: Done. Tests pass. Thanks.
